torch/distributed/_functional_collectives.py

```diff
@@ -0,0 +1,231 @@
+from typing import Any, Tuple, Union, List, cast
+
+import weakref
+import warnings
+
+import torch
+import torch.distributed as dist
+
+from torch._C import _disabled_torch_function_impl
+from torch.utils._pytree import tree_map
+
+import torch.distributed.distributed_c10d as c10d
+"""
+New traceable, functional collectives.
+  compiler: trace these ops with plain-old-data schemas, then choose how to lower them.
+  eager: execute these 'functional' ops which in eager return AsyncCollectiveTensor subclasses,
+         automatically calling .wait() on underlying/hidden async 'work' obj only when fed to
+         a downstream op.
+
+Issues:
+* Where should these ops live? Couldn't `import torch` if putting these ops in existing torch.distributed files
+* Proper support for eager requires inplace ops. We should explore having it as an option for the API.
+"""
+
+"""
+Functional collectives are asynchronous only and we perform implicit stream synchronization
+on behalf of the user.
+
+We use AsyncCollectiveTensor to wrap the result tensor of a collective and it lets us witness
+first usage of the tensor and insert cross stream sync at the right place.
+
+The above are the easy bit, the hard ones are how we match the Work object returned by
+c10d and the tensor AsyncCollectiveTensor wraps. We alloc the tensor inside the collective
+op implementation (see ``clone()`` call in ``_all_reduce``) and then it goes into a voyage
+of wonders and magic through the dispatcher and eventually something comes out of it.
```

Comment:
"and eventually something comes out of it"

would be nice to explain this part actually.  I think what you mean is "and eventually the dispatcher returns a new pyobj Tensor to the caller, but it may not be the same pyobj as the one initially returned by `clone()`"?